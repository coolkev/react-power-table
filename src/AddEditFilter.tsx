import * as React from 'react';
import * as Button from 'react-bootstrap/lib/Button';
import * as Radio from 'react-bootstrap/lib/Radio';
import * as filters from './filters/FilterDefinition';
import { debuglog } from './utils';

/**
 * @internal
 */
export interface AddEditFilterProps {
    filter: filters.FilterDefinition<any>;
    initialOperation: filters.OperationDefinition<any>;
    initialValue?: any;
    onApplyFilter: (filter: filters.AppliedFilter<any>) => void;
}

/**
 * @internal
 */
export interface AddEditFilterState {

    operationKey: string;
    value: any;

}

/**
 * @internal
 */
export class AddEditFilter extends React.PureComponent<AddEditFilterProps, AddEditFilterState> {

    constructor(props: AddEditFilterProps) {
        super(props);

        const value = props.initialValue || props.filter.defaultValue || '';

        this.state = { value, operationKey: props.initialOperation.key };

        this.selectedOperationChange = this.selectedOperationChange.bind(this);
        this.handleFilterValueChange = this.handleFilterValueChange.bind(this);
        this.applyFilter = this.applyFilter.bind(this);

    }

    private selectedOperationChange(e: React.FormEvent<Radio & HTMLInputElement>) {

        const prevOperationKey = this.state.operationKey;
        const operationKey = e.currentTarget.value;
        //const operation = this.props.filter.operations[operationKey];

        const newState: AddEditFilterState = { operationKey, value: this.state.value };

        this.props.filter.onOperationChange(prevOperationKey, newState);
        // if (prevOperationKey === 'between' && operationKey !== prevOperationKey && typeof this.state.value === 'string') {
        //     newState.value = this.state.value.split(' ')[0];
        // }
        this.setState(newState);
    }

    private handleFilterValueChange(value: string) {
        debuglog('AddEditFilter.filterValueChange', value);
        this.setState({
            value,
        });
    }

    private applyFilter() {
        const operation = this.props.filter.operations[this.state.operationKey];

        this.props.onApplyFilter({ filter: this.props.filter, operation, value: this.state.value });
    }
    render() {
        debuglog('AddEditFilter.render', this.props);

        const filter = this.props.filter;

        const { value, operationKey } = this.state;

        const opComponents = Object.keys(filter.operations).map((opKey) => {
            const op = filter.operations[opKey];
            let children: React.ReactChild;
            if (opKey === this.state.operationKey) {

                const SelectedFilterComponent = op.filterComponent || filter.filterComponent;

                if (SelectedFilterComponent) {
                    children = <div style={{ marginLeft: 20 }}><SelectedFilterComponent filter={filter} operation={op} value={value} onValueChange={this.handleFilterValueChange} onEnterKeyPress={this.applyFilter} /></div>;
                }
            }

            return <FilterOperation key={opKey} operation={op} filter={filter} selected={opKey === operationKey} onChange={this.selectedOperationChange}>{children}</FilterOperation>;
        });

        return (
            <div>
                <div style={{ margin: '10px 0' }}><b>{filter.displayName}</b>
                </div>

                {opComponents.map((opComponent) => opComponent)}

                <div style={{ marginTop: 20 }}>
                    <Button bsStyle="primary" bsSize="sm" onClick={this.applyFilter}>Apply Filter</Button>
                </div>

            </div>);
    }
}

/**
 * @internal
 */
export interface FilterOperationProps extends React.Props<any> {
    operation: filters.OperationDefinition;
    filter: filters.FilterDefinition;

    selected: boolean;
    onChange: React.EventHandler<React.FormEvent<Radio & HTMLInputElement>>;
}

/**
 * @internal
 */
export const FilterOperation = (props: FilterOperationProps) => {

    const { operation, filter, onChange, selected, children } = props;

    return (
        <div key={operation.key}>
            <Radio checked={selected} onChange={onChange} value={operation.key}>
                {filter.radioButtonLabel ? filter.radioButtonLabel({ filter, operation }) : operation.displayName}

            </Radio>
            {children}
        </div>
    );

};
