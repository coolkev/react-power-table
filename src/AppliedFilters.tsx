import * as React from 'react';
import { ActionLink } from './components/ActionLink';
import * as filters from './filters';

export interface AppliedFiltersProps {
    appliedFilters: ReadonlyArray<filters.AppliedFilter>;
    removeFilter: (filter: filters.AppliedFilter) => void;
    editFilter: (filter: filters.AppliedFilter) => void;

}

export const AppliedFilters = (props: AppliedFiltersProps) => {
    return (
        <div className="applied-filters small">
            {props.appliedFilters.map((appliedFilter) => <AppliedFilter appliedFilter={appliedFilter} key={appliedFilter.key} onEditFilter={props.editFilter} onRemoveFilter={props.removeFilter} />)}
        </div>
    );
};

export interface AppliedFilterProps {
    appliedFilter: filters.AppliedFilter;
    onRemoveFilter: (filter: filters.AppliedFilter) => void;
    onEditFilter: (filter: filters.AppliedFilter) => void;

}

const defaultAppliedFilterLabelComponent = (props: filters.AppliedFilter<any>) => {
    const appliedLabel = props.operation.appliedLabel || props.filter.appliedLabel || filters.FilterDefinition.defaultAppliedFilterLabel;
    return <span>{appliedLabel(props)}</span>;
};

export const AppliedFilter = ({ appliedFilter, onRemoveFilter, onEditFilter }: AppliedFilterProps) => {

    const handleRemoveFilter = React.useCallback((e: React.SyntheticEvent<any>) => {
        e.preventDefault();
        onRemoveFilter(appliedFilter);
    }, [appliedFilter, onRemoveFilter]);

    const handleEditFilter = React.useCallback(() => {

        onEditFilter(appliedFilter);

    }, [appliedFilter, onEditFilter]);

    const AppliedLabelComponent = appliedFilter.operation.appliedLabelComponent || appliedFilter.filter.appliedLabelComponent || defaultAppliedFilterLabelComponent;

    return (
        <div className="applied-filter well well-sm" style={{ marginBottom: 10 }} key={appliedFilter.filter.fieldName}>
            <button type="button" className="close" aria-label="Remove" onClick={handleRemoveFilter}><span aria-hidden="true">Ã—</span></button>
            <ActionLink onClick={handleEditFilter}><AppliedLabelComponent {...appliedFilter} /></ActionLink>
        </div>
    );

};

