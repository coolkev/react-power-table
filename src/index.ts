export * from './ReactPowerTable';
export * from './Paging';
export * from './Sorting';
export * from './GridFilters';
export * from './filters/';
export * from './Column';
export * from './utils';

import { PowerTableProps } from './ReactPowerTable';
export type ReactPowerTable<TRow = {}, TExtraProps = {}> = React.ComponentClass<PowerTableProps<TRow, TExtraProps>>;
