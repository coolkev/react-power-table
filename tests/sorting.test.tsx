import { mount, render } from 'enzyme';
import * as React from 'react';
import { InternalSortingProps, InternalSortingState, ReactPowerTable, withInternalSorting } from '../src/';
import { defaultColumns, sampledata } from './shared';

const columns = defaultColumns;
const rows = sampledata.slice(0, 5);

describe('withInternalSorting tests',
    () => {

        test('change sort on th click', () => {

            const Table = withInternalSorting(ReactPowerTable);

            const c = <Table columns={columns} rows={rows} keyColumn="number" sorting={{ column: 'number' }} />;
            const component = mount<InternalSortingProps, InternalSortingState<any>>(c);

            expect(component.state('currentSort').column).toEqual('number');

            const th = component.find('th');

            expect(th.at(0).text()).toEqual('number ');
            expect(th.at(1).text()).toEqual('name');
            expect(th.at(0).render()).toMatchSnapshot();
            expect(th.at(1).render()).toMatchSnapshot();

            expect(component.render()).toMatchSnapshot();

            th.at(1).simulate('click');

            expect(component.state('currentSort').column).toEqual('president');

            expect(th.at(1).text()).toEqual('name ');
            expect(th.at(0).render()).toMatchSnapshot();
            expect(th.at(1).render()).toMatchSnapshot();

            expect(component.render()).toMatchSnapshot();

        });

        test('change sort from props', () => {

            const Table = withInternalSorting(ReactPowerTable);

            const c = <Table columns={columns} rows={rows} keyColumn="number" sorting={{ column: 'number' }} />;

            const component = mount(c);
            expect(component.state('currentSort').descending).toBeFalsy();

            component.setProps({ sorting: { column: 'number', descending: true } });
            expect(component.state('currentSort').descending).toEqual(true);

            expect(component.render()).toMatchSnapshot();

        });

        test('change rows from props', () => {

            const Table = withInternalSorting(ReactPowerTable);

            const c = <Table columns={columns} rows={rows} keyColumn="number" sorting={{ column: 'number' }} />;

            const component = mount(c);
            expect(component.find('tbody tr').length).toEqual(5);

            component.setProps({ rows: sampledata.slice(0, 4) });

            expect(component.find('tbody tr').length).toEqual(4);
            expect(component.render()).toMatchSnapshot();

        });

        test('raises sortChanging and sortChanged events', () => {

            const Table = withInternalSorting(ReactPowerTable);

            let sortChangingCalled = false;
            let sortChangedCalled = false;

            const onSortChanging = () => sortChangingCalled = true;
            const onSortChanged = () => sortChangedCalled = true;

            const c = <Table columns={columns} rows={rows} keyColumn="number" sorting={{ column: 'number', onSortChanging, onSortChanged }} />;
            const component = mount(c);

            const th = component.find('th');

            th.at(1).simulate('click');

            expect(sortChangedCalled).toEqual(true);
            expect(sortChangedCalled).toEqual(true);

        });

        test('sorting with custom cellcomponent', () => {

            const columns2 = columns.map(c => ({ ...c }));
            columns2[0].valueComponent = p => <span>{p.children}</span>;
            columns2[0].textAlign = 'right';

            const Table = withInternalSorting(ReactPowerTable);

            const component = render(<Table columns={columns2} rows={rows} keyColumn="number" sorting={{ column: 'number' }} />);

            expect(component).toMatchSnapshot();

        });
    });
