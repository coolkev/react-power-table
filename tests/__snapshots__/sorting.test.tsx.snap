// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withInternalSorting tests change rows from props 1`] = `
<table>
  <thead>
    <tr>
      <th
        style="cursor: pointer; text-align: left; white-space: nowrap;"
      >
        <div
          style="padding-right: 15px; position: relative;"
        >
          <span>
            number
          </span>
           
          <span
            class="sort-asc"
            style="width: 0px; height: 0px; border-left: 5px solid transparent; border-right: 5px solid transparent; border-bottom: 5px solid gray; position: absolute; margin-left: 5px;"
          />
        </div>
      </th>
      <th
        style="cursor: pointer; text-align: left; white-space: nowrap;"
      >
        <div>
          <span>
            name
          </span>
        </div>
      </th>
      <th
        style="cursor: pointer; text-align: left; white-space: nowrap;"
      >
        <div>
          <span>
            party
          </span>
        </div>
      </th>
      <th
        style="cursor: pointer; text-align: left; white-space: nowrap;"
      >
        <div>
          <span>
            birth_year
          </span>
        </div>
      </th>
      <th
        style="cursor: pointer; text-align: left; white-space: nowrap;"
      >
        <div>
          <span>
            death_year
          </span>
        </div>
      </th>
      <th
        style="cursor: pointer; text-align: right; white-space: nowrap; width: 140px;"
      >
        <div
          style="padding-right: 15px;"
        >
          <span>
            took_office
          </span>
        </div>
      </th>
      <th
        style="cursor: pointer; text-align: left; white-space: nowrap;"
      >
        <div>
          <span>
            left_office
          </span>
        </div>
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <div>
          1
        </div>
      </td>
      <td>
        <div>
          George Washington
        </div>
      </td>
      <td>
        <div>
          No Party
        </div>
      </td>
      <td>
        <div>
          1732
        </div>
      </td>
      <td>
        <div>
          1799
        </div>
      </td>
      <td
        style="width: 140px; text-align: right;"
      >
        <div
          style="margin-right: 15px;"
        >
          <div>
            1789-04-30
          </div>
        </div>
      </td>
      <td>
        <div>
          1797-03-04
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div>
          2
        </div>
      </td>
      <td>
        <div>
          John Adams
        </div>
      </td>
      <td>
        <div>
          Federalist
        </div>
      </td>
      <td>
        <div>
          1735
        </div>
      </td>
      <td>
        <div>
          1826
        </div>
      </td>
      <td
        style="width: 140px; text-align: right;"
      >
        <div
          style="margin-right: 15px;"
        >
          <div>
            1797-03-04
          </div>
        </div>
      </td>
      <td>
        <div>
          1801-03-04
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div>
          3
        </div>
      </td>
      <td>
        <div>
          Thomas Jefferson
        </div>
      </td>
      <td>
        <div>
          Democratic-Republican
        </div>
      </td>
      <td>
        <div>
          1743
        </div>
      </td>
      <td>
        <div>
          1826
        </div>
      </td>
      <td
        style="width: 140px; text-align: right;"
      >
        <div
          style="margin-right: 15px;"
        >
          <div>
            1801-03-04
          </div>
        </div>
      </td>
      <td>
        <div>
          1809-03-04
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div>
          4
        </div>
      </td>
      <td>
        <div>
          James Madison
        </div>
      </td>
      <td>
        <div>
          Democratic-Republican
        </div>
      </td>
      <td>
        <div>
          1751
        </div>
      </td>
      <td>
        <div>
          1836
        </div>
      </td>
      <td
        style="width: 140px; text-align: right;"
      >
        <div
          style="margin-right: 15px;"
        >
          <div>
            1809-03-04
          </div>
        </div>
      </td>
      <td>
        <div>
          1817-03-04
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`withInternalSorting tests change sort from props 1`] = `
<table>
  <thead>
    <tr>
      <th
        style="cursor: pointer; text-align: left; white-space: nowrap;"
      >
        <div
          style="padding-right: 15px; position: relative;"
        >
          <span>
            number
          </span>
           
          <span
            class="sort-desc"
            style="width: 0px; height: 0px; border-left: 5px solid transparent; border-right: 5px solid transparent; position: absolute; margin-left: 5px; border-top: 5px solid gray;"
          />
        </div>
      </th>
      <th
        style="cursor: pointer; text-align: left; white-space: nowrap;"
      >
        <div>
          <span>
            name
          </span>
        </div>
      </th>
      <th
        style="cursor: pointer; text-align: left; white-space: nowrap;"
      >
        <div>
          <span>
            party
          </span>
        </div>
      </th>
      <th
        style="cursor: pointer; text-align: left; white-space: nowrap;"
      >
        <div>
          <span>
            birth_year
          </span>
        </div>
      </th>
      <th
        style="cursor: pointer; text-align: left; white-space: nowrap;"
      >
        <div>
          <span>
            death_year
          </span>
        </div>
      </th>
      <th
        style="cursor: pointer; text-align: right; white-space: nowrap; width: 140px;"
      >
        <div
          style="padding-right: 15px;"
        >
          <span>
            took_office
          </span>
        </div>
      </th>
      <th
        style="cursor: pointer; text-align: left; white-space: nowrap;"
      >
        <div>
          <span>
            left_office
          </span>
        </div>
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <div>
          5
        </div>
      </td>
      <td>
        <div>
          James Monroe
        </div>
      </td>
      <td>
        <div>
          Democratic-Republican
        </div>
      </td>
      <td>
        <div>
          1758
        </div>
      </td>
      <td>
        <div>
          1831
        </div>
      </td>
      <td
        style="width: 140px; text-align: right;"
      >
        <div
          style="margin-right: 15px;"
        >
          <div>
            1817-03-04
          </div>
        </div>
      </td>
      <td>
        <div>
          1825-03-04
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div>
          4
        </div>
      </td>
      <td>
        <div>
          James Madison
        </div>
      </td>
      <td>
        <div>
          Democratic-Republican
        </div>
      </td>
      <td>
        <div>
          1751
        </div>
      </td>
      <td>
        <div>
          1836
        </div>
      </td>
      <td
        style="width: 140px; text-align: right;"
      >
        <div
          style="margin-right: 15px;"
        >
          <div>
            1809-03-04
          </div>
        </div>
      </td>
      <td>
        <div>
          1817-03-04
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div>
          3
        </div>
      </td>
      <td>
        <div>
          Thomas Jefferson
        </div>
      </td>
      <td>
        <div>
          Democratic-Republican
        </div>
      </td>
      <td>
        <div>
          1743
        </div>
      </td>
      <td>
        <div>
          1826
        </div>
      </td>
      <td
        style="width: 140px; text-align: right;"
      >
        <div
          style="margin-right: 15px;"
        >
          <div>
            1801-03-04
          </div>
        </div>
      </td>
      <td>
        <div>
          1809-03-04
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div>
          2
        </div>
      </td>
      <td>
        <div>
          John Adams
        </div>
      </td>
      <td>
        <div>
          Federalist
        </div>
      </td>
      <td>
        <div>
          1735
        </div>
      </td>
      <td>
        <div>
          1826
        </div>
      </td>
      <td
        style="width: 140px; text-align: right;"
      >
        <div
          style="margin-right: 15px;"
        >
          <div>
            1797-03-04
          </div>
        </div>
      </td>
      <td>
        <div>
          1801-03-04
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div>
          1
        </div>
      </td>
      <td>
        <div>
          George Washington
        </div>
      </td>
      <td>
        <div>
          No Party
        </div>
      </td>
      <td>
        <div>
          1732
        </div>
      </td>
      <td>
        <div>
          1799
        </div>
      </td>
      <td
        style="width: 140px; text-align: right;"
      >
        <div
          style="margin-right: 15px;"
        >
          <div>
            1789-04-30
          </div>
        </div>
      </td>
      <td>
        <div>
          1797-03-04
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`withInternalSorting tests change sort on th click 1`] = `
<th
  style="cursor: pointer; text-align: left; white-space: nowrap;"
>
  <div
    style="padding-right: 15px; position: relative;"
  >
    <span>
      number
    </span>
     
    <span
      class="sort-asc"
      style="width: 0px; height: 0px; border-left: 5px solid transparent; border-right: 5px solid transparent; border-bottom: 5px solid gray; position: absolute; margin-left: 5px;"
    />
  </div>
</th>
`;

exports[`withInternalSorting tests change sort on th click 2`] = `
<th
  style="cursor: pointer; text-align: left; white-space: nowrap;"
>
  <div>
    <span>
      name
    </span>
  </div>
</th>
`;

exports[`withInternalSorting tests change sort on th click 3`] = `
<table>
  <thead>
    <tr>
      <th
        style="cursor: pointer; text-align: left; white-space: nowrap;"
      >
        <div
          style="padding-right: 15px; position: relative;"
        >
          <span>
            number
          </span>
           
          <span
            class="sort-asc"
            style="width: 0px; height: 0px; border-left: 5px solid transparent; border-right: 5px solid transparent; border-bottom: 5px solid gray; position: absolute; margin-left: 5px;"
          />
        </div>
      </th>
      <th
        style="cursor: pointer; text-align: left; white-space: nowrap;"
      >
        <div>
          <span>
            name
          </span>
        </div>
      </th>
      <th
        style="cursor: pointer; text-align: left; white-space: nowrap;"
      >
        <div>
          <span>
            party
          </span>
        </div>
      </th>
      <th
        style="cursor: pointer; text-align: left; white-space: nowrap;"
      >
        <div>
          <span>
            birth_year
          </span>
        </div>
      </th>
      <th
        style="cursor: pointer; text-align: left; white-space: nowrap;"
      >
        <div>
          <span>
            death_year
          </span>
        </div>
      </th>
      <th
        style="cursor: pointer; text-align: right; white-space: nowrap; width: 140px;"
      >
        <div
          style="padding-right: 15px;"
        >
          <span>
            took_office
          </span>
        </div>
      </th>
      <th
        style="cursor: pointer; text-align: left; white-space: nowrap;"
      >
        <div>
          <span>
            left_office
          </span>
        </div>
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <div>
          1
        </div>
      </td>
      <td>
        <div>
          George Washington
        </div>
      </td>
      <td>
        <div>
          No Party
        </div>
      </td>
      <td>
        <div>
          1732
        </div>
      </td>
      <td>
        <div>
          1799
        </div>
      </td>
      <td
        style="width: 140px; text-align: right;"
      >
        <div
          style="margin-right: 15px;"
        >
          <div>
            1789-04-30
          </div>
        </div>
      </td>
      <td>
        <div>
          1797-03-04
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div>
          2
        </div>
      </td>
      <td>
        <div>
          John Adams
        </div>
      </td>
      <td>
        <div>
          Federalist
        </div>
      </td>
      <td>
        <div>
          1735
        </div>
      </td>
      <td>
        <div>
          1826
        </div>
      </td>
      <td
        style="width: 140px; text-align: right;"
      >
        <div
          style="margin-right: 15px;"
        >
          <div>
            1797-03-04
          </div>
        </div>
      </td>
      <td>
        <div>
          1801-03-04
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div>
          3
        </div>
      </td>
      <td>
        <div>
          Thomas Jefferson
        </div>
      </td>
      <td>
        <div>
          Democratic-Republican
        </div>
      </td>
      <td>
        <div>
          1743
        </div>
      </td>
      <td>
        <div>
          1826
        </div>
      </td>
      <td
        style="width: 140px; text-align: right;"
      >
        <div
          style="margin-right: 15px;"
        >
          <div>
            1801-03-04
          </div>
        </div>
      </td>
      <td>
        <div>
          1809-03-04
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div>
          4
        </div>
      </td>
      <td>
        <div>
          James Madison
        </div>
      </td>
      <td>
        <div>
          Democratic-Republican
        </div>
      </td>
      <td>
        <div>
          1751
        </div>
      </td>
      <td>
        <div>
          1836
        </div>
      </td>
      <td
        style="width: 140px; text-align: right;"
      >
        <div
          style="margin-right: 15px;"
        >
          <div>
            1809-03-04
          </div>
        </div>
      </td>
      <td>
        <div>
          1817-03-04
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div>
          5
        </div>
      </td>
      <td>
        <div>
          James Monroe
        </div>
      </td>
      <td>
        <div>
          Democratic-Republican
        </div>
      </td>
      <td>
        <div>
          1758
        </div>
      </td>
      <td>
        <div>
          1831
        </div>
      </td>
      <td
        style="width: 140px; text-align: right;"
      >
        <div
          style="margin-right: 15px;"
        >
          <div>
            1817-03-04
          </div>
        </div>
      </td>
      <td>
        <div>
          1825-03-04
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`withInternalSorting tests change sort on th click 4`] = `
<th
  style="cursor: pointer; text-align: left; white-space: nowrap;"
>
  <div
    style="padding-right: 15px;"
  >
    <span>
      number
    </span>
  </div>
</th>
`;

exports[`withInternalSorting tests change sort on th click 5`] = `
<th
  style="cursor: pointer; text-align: left; white-space: nowrap;"
>
  <div
    style="padding-right: 15px; position: relative;"
  >
    <span>
      name
    </span>
     
    <span
      class="sort-asc"
      style="width: 0px; height: 0px; border-left: 5px solid transparent; border-right: 5px solid transparent; border-bottom: 5px solid gray; position: absolute; margin-left: 5px;"
    />
  </div>
</th>
`;

exports[`withInternalSorting tests change sort on th click 6`] = `
<table>
  <thead>
    <tr>
      <th
        style="cursor: pointer; text-align: left; white-space: nowrap;"
      >
        <div
          style="padding-right: 15px;"
        >
          <span>
            number
          </span>
        </div>
      </th>
      <th
        style="cursor: pointer; text-align: left; white-space: nowrap;"
      >
        <div
          style="padding-right: 15px; position: relative;"
        >
          <span>
            name
          </span>
           
          <span
            class="sort-asc"
            style="width: 0px; height: 0px; border-left: 5px solid transparent; border-right: 5px solid transparent; border-bottom: 5px solid gray; position: absolute; margin-left: 5px;"
          />
        </div>
      </th>
      <th
        style="cursor: pointer; text-align: left; white-space: nowrap;"
      >
        <div>
          <span>
            party
          </span>
        </div>
      </th>
      <th
        style="cursor: pointer; text-align: left; white-space: nowrap;"
      >
        <div>
          <span>
            birth_year
          </span>
        </div>
      </th>
      <th
        style="cursor: pointer; text-align: left; white-space: nowrap;"
      >
        <div>
          <span>
            death_year
          </span>
        </div>
      </th>
      <th
        style="cursor: pointer; text-align: right; white-space: nowrap; width: 140px;"
      >
        <div
          style="padding-right: 15px;"
        >
          <span>
            took_office
          </span>
        </div>
      </th>
      <th
        style="cursor: pointer; text-align: left; white-space: nowrap;"
      >
        <div>
          <span>
            left_office
          </span>
        </div>
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <div>
          1
        </div>
      </td>
      <td>
        <div>
          George Washington
        </div>
      </td>
      <td>
        <div>
          No Party
        </div>
      </td>
      <td>
        <div>
          1732
        </div>
      </td>
      <td>
        <div>
          1799
        </div>
      </td>
      <td
        style="width: 140px; text-align: right;"
      >
        <div
          style="margin-right: 15px;"
        >
          <div>
            1789-04-30
          </div>
        </div>
      </td>
      <td>
        <div>
          1797-03-04
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div>
          4
        </div>
      </td>
      <td>
        <div>
          James Madison
        </div>
      </td>
      <td>
        <div>
          Democratic-Republican
        </div>
      </td>
      <td>
        <div>
          1751
        </div>
      </td>
      <td>
        <div>
          1836
        </div>
      </td>
      <td
        style="width: 140px; text-align: right;"
      >
        <div
          style="margin-right: 15px;"
        >
          <div>
            1809-03-04
          </div>
        </div>
      </td>
      <td>
        <div>
          1817-03-04
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div>
          5
        </div>
      </td>
      <td>
        <div>
          James Monroe
        </div>
      </td>
      <td>
        <div>
          Democratic-Republican
        </div>
      </td>
      <td>
        <div>
          1758
        </div>
      </td>
      <td>
        <div>
          1831
        </div>
      </td>
      <td
        style="width: 140px; text-align: right;"
      >
        <div
          style="margin-right: 15px;"
        >
          <div>
            1817-03-04
          </div>
        </div>
      </td>
      <td>
        <div>
          1825-03-04
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div>
          2
        </div>
      </td>
      <td>
        <div>
          John Adams
        </div>
      </td>
      <td>
        <div>
          Federalist
        </div>
      </td>
      <td>
        <div>
          1735
        </div>
      </td>
      <td>
        <div>
          1826
        </div>
      </td>
      <td
        style="width: 140px; text-align: right;"
      >
        <div
          style="margin-right: 15px;"
        >
          <div>
            1797-03-04
          </div>
        </div>
      </td>
      <td>
        <div>
          1801-03-04
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div>
          3
        </div>
      </td>
      <td>
        <div>
          Thomas Jefferson
        </div>
      </td>
      <td>
        <div>
          Democratic-Republican
        </div>
      </td>
      <td>
        <div>
          1743
        </div>
      </td>
      <td>
        <div>
          1826
        </div>
      </td>
      <td
        style="width: 140px; text-align: right;"
      >
        <div
          style="margin-right: 15px;"
        >
          <div>
            1801-03-04
          </div>
        </div>
      </td>
      <td>
        <div>
          1809-03-04
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`withInternalSorting tests sorting with custom cellcomponent 1`] = `
<table>
  <thead>
    <tr>
      <th
        style="cursor:pointer;text-align:right;white-space:nowrap;"
      >
        <div
          style="padding-right:15px;position:relative;"
        >
          <span>
            number
          </span>
           
          <span
            class="sort-asc"
            style="width:0;height:0;border-left:5px solid transparent;border-right:5px solid transparent;border-bottom:5px solid gray;position:absolute;top:calc(50% - 3px);margin-left:5px;"
          />
        </div>
      </th>
      <th
        style="cursor:pointer;text-align:left;white-space:nowrap;"
      >
        <div>
          <span>
            name
          </span>
        </div>
      </th>
      <th
        style="cursor:pointer;text-align:left;white-space:nowrap;"
      >
        <div>
          <span>
            party
          </span>
        </div>
      </th>
      <th
        style="cursor:pointer;text-align:left;white-space:nowrap;"
      >
        <div>
          <span>
            birth_year
          </span>
        </div>
      </th>
      <th
        style="cursor:pointer;text-align:left;white-space:nowrap;"
      >
        <div>
          <span>
            death_year
          </span>
        </div>
      </th>
      <th
        style="cursor:pointer;text-align:right;white-space:nowrap;width:140px;"
      >
        <div
          style="padding-right:15px;"
        >
          <span>
            took_office
          </span>
        </div>
      </th>
      <th
        style="cursor:pointer;text-align:left;white-space:nowrap;"
      >
        <div>
          <span>
            left_office
          </span>
        </div>
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td
        style="text-align:right;"
      >
        <div
          style="margin-right:15px;"
        >
          <span />
        </div>
      </td>
      <td>
        <div>
          George Washington
        </div>
      </td>
      <td>
        <div>
          No Party
        </div>
      </td>
      <td>
        <div>
          1732
        </div>
      </td>
      <td>
        <div>
          1799
        </div>
      </td>
      <td
        style="width:140px;text-align:right;"
      >
        <div
          style="margin-right:15px;"
        >
          <div>
            1789-04-30
          </div>
        </div>
      </td>
      <td>
        <div>
          1797-03-04
        </div>
      </td>
    </tr>
    <tr>
      <td
        style="text-align:right;"
      >
        <div
          style="margin-right:15px;"
        >
          <span />
        </div>
      </td>
      <td>
        <div>
          John Adams
        </div>
      </td>
      <td>
        <div>
          Federalist
        </div>
      </td>
      <td>
        <div>
          1735
        </div>
      </td>
      <td>
        <div>
          1826
        </div>
      </td>
      <td
        style="width:140px;text-align:right;"
      >
        <div
          style="margin-right:15px;"
        >
          <div>
            1797-03-04
          </div>
        </div>
      </td>
      <td>
        <div>
          1801-03-04
        </div>
      </td>
    </tr>
    <tr>
      <td
        style="text-align:right;"
      >
        <div
          style="margin-right:15px;"
        >
          <span />
        </div>
      </td>
      <td>
        <div>
          Thomas Jefferson
        </div>
      </td>
      <td>
        <div>
          Democratic-Republican
        </div>
      </td>
      <td>
        <div>
          1743
        </div>
      </td>
      <td>
        <div>
          1826
        </div>
      </td>
      <td
        style="width:140px;text-align:right;"
      >
        <div
          style="margin-right:15px;"
        >
          <div>
            1801-03-04
          </div>
        </div>
      </td>
      <td>
        <div>
          1809-03-04
        </div>
      </td>
    </tr>
    <tr>
      <td
        style="text-align:right;"
      >
        <div
          style="margin-right:15px;"
        >
          <span />
        </div>
      </td>
      <td>
        <div>
          James Madison
        </div>
      </td>
      <td>
        <div>
          Democratic-Republican
        </div>
      </td>
      <td>
        <div>
          1751
        </div>
      </td>
      <td>
        <div>
          1836
        </div>
      </td>
      <td
        style="width:140px;text-align:right;"
      >
        <div
          style="margin-right:15px;"
        >
          <div>
            1809-03-04
          </div>
        </div>
      </td>
      <td>
        <div>
          1817-03-04
        </div>
      </td>
    </tr>
    <tr>
      <td
        style="text-align:right;"
      >
        <div
          style="margin-right:15px;"
        >
          <span />
        </div>
      </td>
      <td>
        <div>
          James Monroe
        </div>
      </td>
      <td>
        <div>
          Democratic-Republican
        </div>
      </td>
      <td>
        <div>
          1758
        </div>
      </td>
      <td>
        <div>
          1831
        </div>
      </td>
      <td
        style="width:140px;text-align:right;"
      >
        <div
          style="margin-right:15px;"
        >
          <div>
            1817-03-04
          </div>
        </div>
      </td>
      <td>
        <div>
          1825-03-04
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;
